# TFIDF Vectorization 

TF-IDF (Term Frequency–Inverse Document Frequency) is a text vectorization technique that converts textual data into numerical feature vectors by assigning weights to words based on their importance in a document relative to the entire corpus. Words that occur frequently in a document but rarely across other documents receive higher weights, while commonly occurring words receive lower weights.

In [1]:
import pandas as pd
import numpy as np
import itertools
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer, HashingVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn import metrics
import matplotlib.pyplot as plt

In [4]:
dataset = pd.read_csv("news.csv").drop(columns=["Unnamed: 0"])

In [5]:
dataset

,title,text,label
0,You Can Smell Hillary’s Fear,"Daniel Greenfield, a Shillman Journalism Fello...",FAKE
1,Watch The Exact Moment Paul Ryan Committed Pol...,Google Pinterest Digg Linkedin Reddit Stumbleu...,FAKE
2,Kerry to go to Paris in gesture of sympathy,U.S. Secretary of State John F. Kerry said Mon...,REAL
3,Bernie supporters on Twitter erupt in anger ag...,"— Kaydee King (@KaydeeKing) November 9, 2016 T...",FAKE
4,The Battle of New York: Why This Primary Matters,It's primary day in New York and front-runners...,REAL
...,...,...,...
6330,State Department says it can't find emails fro...,The State Department told the Republican Natio...,REAL
6331,The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...,The ‘P’ in PBS Should Stand for ‘Plutocratic’ ...,FAKE
6332,Anti-Trump Protesters Are Tools of the Oligarc...,Anti-Trump Protesters Are Tools of the Oligar...,FAKE
6333,"In Ethiopia, Obama seeks progress on peace, se...","ADDIS ABABA, Ethiopia —President Obama convene...",REAL


In [6]:
y=dataset["label"]

In [7]:
X_train, X_test, y_train, y_test = train_test_split(dataset['text'], y, test_size=0.33, random_state=53)

In [8]:
tfidf_vectorizer  = TfidfVectorizer(stop_words='english')
tfidf_train = tfidf_vectorizer.fit_transform(X_train)
print(tfidf_train)
tfidf_test = tfidf_vectorizer.transform(X_test)

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 1119820 stored elements and shape (4244, 56922)>
  Coords	Values
  (1, 36831)	0.10772488937433063
  (1, 47506)	0.1453906087746004
  (1, 38823)	0.06048803110658641
  (1, 25684)	0.10304209221897535
  (1, 21568)	0.10079209195663975
  (1, 36087)	0.1264867985438989
  (1, 16814)	0.10404977746548134
  (1, 49203)	0.16727408617843761
  (1, 25686)	0.13550453594288978
  (1, 15927)	0.4973826512693339
  (1, 29531)	0.1454406205718244
  (1, 8399)	0.08729542880625331
  (1, 42534)	0.06081715886809214
  (1, 21881)	0.21271688045815967
  (1, 35256)	0.12628385718450852
  (1, 35783)	0.3090269081882796
  (1, 51896)	0.11596517664605863
  (1, 44520)	0.4973826512693339
  (1, 15924)	0.3479045460649077
  (1, 50628)	0.061296988343109565
  (1, 54177)	0.1378262914471113
  (1, 12105)	0.1500806646147686
  (1, 42470)	0.07711040274149523
  (2, 9445)	0.22180388050235564
  (2, 8779)	0.19767342113068423
  :	:
  (4243, 30464)	0.032161542218569955
  (4243, 35836)	

fit() - creates library for unique words in the document.
transform() - converted into vector . (weights) not counts of words.

each document becomes a vector, vocabularies became columns. most values are 0 because each document contains only a few words. 
A sparse matrix is simply a matrix in which most values are zero.

 (1, 36831)	0.10772488937433063 --> 1st doc, 36831 col, 0.1077 tf-idf weight of that word.

In [10]:
len(tfidf_vectorizer.get_feature_names_out())

56922

In [13]:
clf = MultinomialNB()

clf.fit(tfidf_train, y_train)
pred = clf.predict(tfidf_test)
score = metrics.accuracy_score(y_test, pred)
print("accuracy:   %0.3f" % score)

cm = metrics.confusion_matrix(y_test, pred, labels=['FAKE', 'REAL'])

accuracy:   0.857


In [15]:
from sklearn.metrics import classification_report

report=classification_report(y_test, pred)


In [17]:
print(report)

              precision    recall  f1-score   support

        FAKE       0.96      0.73      0.83      1008
        REAL       0.80      0.97      0.88      1083

    accuracy                           0.86      2091
   macro avg       0.88      0.85      0.85      2091
weighted avg       0.88      0.86      0.85      2091



In [31]:
dataset["text"][2]

'U.S. Secretary of State John F. Kerry said Monday that he will stop in Paris later this week, amid criticism that no top American officials attended Sunday’s unity march against terrorism.\n\nKerry said he expects to arrive in Paris Thursday evening, as he heads home after a week abroad. He said he will fly to France at the conclusion of a series of meetings scheduled for Thursday in Sofia, Bulgaria. He plans to meet the next day with Foreign Minister Laurent Fabius and President Francois Hollande, then return to Washington.\n\nThe visit by Kerry, who has family and childhood ties to the country and speaks fluent French, could address some of the criticism that the United States snubbed France in its darkest hour in many years.\n\nThe French press on Monday was filled with questions about why neither President Obama nor Kerry attended Sunday’s march, as about 40 leaders of other nations did. Obama was said to have stayed away because his own security needs can be taxing on a country, 

In [32]:
tfidf_train[[2]]

<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 9 stored elements and shape (1, 56922)>

In [33]:
tfidf_vectorizer  = TfidfVectorizer(stop_words='english')
tfidf_train = tfidf_vectorizer.fit_transform(X_train)

In [34]:
X_train[[2]]

2    U.S. Secretary of State John F. Kerry said Mon...
Name: text, dtype: object

In [35]:
clf.predict(tfidf_train[[2]])

array(['REAL'], dtype='<U4')